src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs

```diff
@@ -0,0 +1,1182 @@
+            {
+                _buffer.Grow();
+            }
+            _buffer.Commit(bytesWritten);
+        }
+
+        private void BufferLiteralHeaderWithStaticNameReference(int nameIndex, string value)
+        {
+            int bytesWritten;
+            while (!QPackEncoder.EncodeLiteralHeaderFieldWithStaticNameReference(nameIndex, value, _buffer.AvailableSpan, out bytesWritten))
+            {
+                _buffer.Grow();
+            }
+            _buffer.Commit(bytesWritten);
+        }
+
+        private void BufferLiteralHeaderWithoutNameReference(string name, ReadOnlySpan<string> values, string separator)
+        {
+            int bytesWritten;
+            while (!QPackEncoder.EncodeLiteralHeaderFieldWithoutNameReference(name, values, separator, _buffer.AvailableSpan, out bytesWritten))
+            {
+                _buffer.Grow();
+            }
+            _buffer.Commit(bytesWritten);
+        }
+
+        private void BufferLiteralHeaderWithoutNameReference(string name, string value)
+        {
+            int bytesWritten;
+            while (!QPackEncoder.EncodeLiteralHeaderFieldWithoutNameReference(name, value, _buffer.AvailableSpan, out bytesWritten))
+            {
+                _buffer.Grow();
+            }
+            _buffer.Commit(bytesWritten);
+        }
+
+        private void BufferLiteralHeaderValues(ReadOnlySpan<string> values, string separator)
+        {
+            int bytesWritten;
+            while (!QPackEncoder.EncodeValueString(values, separator, _buffer.AvailableSpan, out bytesWritten))
+            {
+                _buffer.Grow();
+            }
+            _buffer.Commit(bytesWritten);
+        }
+
+        private void BufferFrameEnvelope(Http3FrameType frameType, long payloadLength)
+        {
+            int bytesWritten;
+            while (!Http3Frame.TryWriteFrameEnvelope(frameType, payloadLength, _buffer.AvailableSpan, out bytesWritten))
+            {
+                _buffer.Grow();
+            }
+            _buffer.Commit(bytesWritten);
+        }
+
+        private void BufferBytes(ReadOnlySpan<byte> span)
+        {
+            _buffer.EnsureAvailableSpace(span.Length);
+            span.CopyTo(_buffer.AvailableSpan);
+            _buffer.Commit(span.Length);
+        }
+
+        private async ValueTask<(Http3FrameType? frameType, long payloadLength)> ReadFrameEnvelopeAsync(CancellationToken cancellationToken)
+        {
+            long frameType, payloadLength;
+            int bytesRead;
+
+            while (true)
+            {
+                while (!Http3Frame.TryReadIntegerPair(_buffer.ActiveSpan, out frameType, out payloadLength, out bytesRead))
+                {
+                    _buffer.EnsureAvailableSpace(2);
+                    bytesRead = await _stream.ReadAsync(_buffer.AvailableMemory, cancellationToken).ConfigureAwait(false);
```

Comment:
These small reads don't seem like they would be good for perf. Do you think it may be better to preemptively allocate some amount in the buffer s.t. future calls that use the payload length already have it?

Reply:
`AavailableMemory` will typically be much larger here. `2` is just the minimum that we need for two integers.